### PR TITLE
Fix onoff test program

### PR DESCRIPTION
**Pull request by pillip8282**

### CODE REVIEW

File: os/run_gdb.sh [hunk 1]
```diff
@@ -0,0 +1 @@
+arm-none-eabi-gdb ../build/output/bin/tinyara -x gdb_commands
```
Comment (jeongchanKim) [inline os/run_gdb.sh:1]: I think this file and the below (`run_openocd.sh`) are not needed. Please remove them.

File: os/drivers/wireless/virtual/vdev_packet.c [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+	VWIFI_LOG("%p, len(%d)\n", buf, len);
+	// print ethernet header
+	int next = 0, sum = 0;
+	int res = vwifi_handle_eth(buf, 0, &next);
+	if (res < 0) {
+		return;
+	}
+	sum += next;
+
+	res =vwifi_handle_ip(buf + sum, 0, &next);
```
Comment (jeongchanKim) [inline os/drivers/wireless/virtual/vdev_packet.c:342]: Please add space after `=`.

File: os/drivers/wireless/virtual/vdev_handler.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#ifndef __VWIFI_HANDLER_H__
```
Comment (jeongchanKim) [inline os/drivers/wireless/virtual/vdev_handler.h:1]: Please add license.

File: os/drivers/wireless/virtual/vdev_handler.c [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+void _generate_disconnect(int argc, char *argv[])
+{
+	sleep(3);
+	vwifi_callback_handler(VWIFI_EVT_STA_DISCONNECTED);
+}
+
+int vwifi_create_event(struct vwifi_evt *event)
+{
+	int res = task_create("vwifi_evt", 100, 1024, event->func, NULL);
```
Comment (jeongchanKim) [inline os/drivers/wireless/virtual/vdev_handler.c:91]: I think we'd better use `kernel_thread` instead of `task_create`, because this is in kernel side.

File: os/drivers/wireless/virtual/vdev.c [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+ * Callback
+ */
+static inline int _create_message(struct vwifi_msg **vmsg, struct vwifi_req *req)
+{
+	struct vwifi_msg *tmsg = (struct vwifi_msg *)malloc(sizeof(struct vwifi_msg));
+	if (!tmsg) {
+		VWIFI_ERROR(0);
+		return -1;
+	}
+	sem_t *signal = (sem_t *)malloc(sizeof(sem_t));
```
Comment (jeongchanKim) [inline os/drivers/wireless/virtual/vdev.c:90]: Please use `kmm_malloc` instead of `malloc`.

File: os/drivers/wireless/virtual/vdev.c [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+	if (!signal) {
+		free(tmsg);
+		VWIFI_ERROR(0);
+		return -1;
+	}
+	sem_init(signal, 0, 0);
+
```
Comment (jeongchanKim) [inline os/drivers/wireless/virtual/vdev.c:97]: Please use `kmm_free` instead of `free`.

File: os/drivers/lwnl/lwnl.c [hunk 7]
```diff
@@ -252,6 +239,17 @@ static int lwnl_poll(FAR struct file *filep, FAR struct pollfd *fds, bool setup)
 
 	/* Are we setting up the poll? Or tearing it down? */
 	if (setup) {
+		/*  Check if any requested events are already in effect */
+		if (fds->events & POLLIN) {
+			int nready = lwnl_check_queue(filep);
+			if (nready > 0) {
+				fds->revents |= (fds->events & POLLIN);
+				sem_post(fds->sem);
+				LWNLDEV_UNLOCK(upper);
```
Comment (jeongchanKim) [inline os/drivers/lwnl/lwnl.c:248]: Is it okay to unlock after `sem_post`? Because, sometimes `sem_post` makes context switching directly before calling unlock.

File: os/drivers/lwnl/lwnl.c [hunk 8]
```diff
@@ -274,6 +272,18 @@ static int lwnl_poll(FAR struct file *filep, FAR struct pollfd *fds, bool setup)
...
 		}
+
+		/* Call lwnl_pollscan again: there could have been events between
+		   the last scan (without us on the list) and putting us on the list! */
+		if (fds->events & POLLIN) {
+			int nready = lwnl_check_queue(filep);
+			if (nready > 0) {
+				fds->revents |= (fds->events & POLLIN);
+				sem_post(fds->sem);
+				LWNLDEV_UNLOCK(upper);
```
Comment (jeongchanKim) [inline os/drivers/lwnl/lwnl.c:283]: Same as above.

File: os/drivers/lwnl/lwnl.c [hunk 9]
```diff
@@ -349,11 +347,11 @@ int lwnl_postmsg(lwnl_cb_status evttype, void *buffer)
 		return -1;
 	}
 
+	LWNLDEV_LOCK(g_lwnl_upper);
 	int res = lwnl_add_event(evttype, buffer);
 	if (res < 0) {
 		return -1;
```
Comment (jeongchanKim) [inline os/drivers/lwnl/lwnl.c:353]: Is it okay not to unlock the `g_lwnl_upper` before return?

File: os/drivers/lwnl/lwnl.c [hunk 10]
```diff
@@ -363,7 +361,7 @@ int lwnl_postmsg(lwnl_cb_status evttype, void *buffer)
 			}
 		}
 	}
-
+	LWNLDEV_UNLOCK(g_lwnl_upper);
```
Comment (jeongchanKim) [inline os/drivers/lwnl/lwnl.c:364]: `sem_post` in line 360 sometimes can make context switching. In this case, there is no unlock the `g_lwnl_upper`.
Is it okay?

File: apps/examples/wifi_manager_sample/wm_test_on_off.c [hunk 11]
```diff
@@ -152,23 +154,33 @@ static void print_wifi_ap_profile(wifi_manager_ap_config_s *config, char *title)
...
 
-static void run_init(void *arg)
+static int run_init(void *arg)
 {
 	wifi_manager_result_e res = wifi_manager_init(&g_wifi_callbacks);
 	if (res != WIFI_MANAGER_SUCCESS) {
-		WM_ERROR(res);
-		return;
+		WO_ERROR(res);
+		return -1;
```
Comment (amogh-samsung) [inline apps/examples/wifi_manager_sample/wm_test_on_off.c:183]: Preferable to return error code MACRO now since `error.h` has been included?

File: apps/examples/wifi_manager_sample/wm_test_on_off.c [hunk 12]
```diff
@@ -21,12 +21,17 @@
 #include <string.h>
 #include <unistd.h>
 #include <sys/types.h>
+#include <semaphore.h>
+#include <errno.h>
```
Comment (amogh-samsung) [inline apps/examples/wifi_manager_sample/wm_test_on_off.c:25]: Why is this needed?

File: apps/examples/wifi_manager_sample/wm_test_on_off.c (hunk 11 above)
Comment (amogh-samsung) [reply apps/examples/wifi_manager_sample/wm_test_on_off.c:183]: Please take a look at comment on Line 25 too

File: build/configs/rtl8721csm/nettest/Make.defs [hunk 13]
```diff
@@ -0,0 +1,215 @@
+###########################################################################
+#
+# Copyright 2020 Samsung Electronics All Rights Reserved.
```
Comment (amogh-samsung) [inline build/configs/rtl8721csm/nettest/Make.defs:3]: We'd have to change the year of this file signature now I think.

File: os/drivers/wireless/virtual/Make.defs [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+# language governing permissions and limitations under the License.
+#
+###########################################################################
+
+ifeq ($(CONFIG_VIRTUAL_WLAN),y)
+
+#Driver API
+CSRCS += vdev.c vdev_handler.c vdev_packet.c
+
+
```
Comment (deepaksrma) [inline os/drivers/wireless/virtual/Make.defs:24]: please remove extra lines

File: os/net/netmgr/netdev_mgr_internal.c [hunk 15]
```diff
@@ -223,6 +212,13 @@ int nm_ifup(struct netdev *dev)
 		}
 	}
 
+	ret = ((struct netdev_ops *)(dev->ops))->ifup(dev);
+	if (ret < 0) {
+		ndbg("fail to up network interface\n");
+		return -1;
+	}
+
+
```
Comment (deepaksrma) [inline os/net/netmgr/netdev_mgr_internal.c:221]: extra line

File: build/configs/rtl8721csm/nettest/Make.defs [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+AFLAGS = $(CFLAGS) -D__ASSEMBLY__
+
+##########################################################################################################################################
+CFLAGS += -march=armv8-m.main+dsp 
+CFLAGS += -mthumb 
+CFLAGS += -mcmse 
+#CFLAGS += -mfloat-abi=hard
+#CFLAGS += -mfpu=fpv5-sp-d16
+
+
```
Comment (deepaksrma) [inline build/configs/rtl8721csm/nettest/Make.defs:161]: please remove extra lines

File: os/drivers/wireless/virtual/Make.defs (hunk 14 above)
Comment (deepaksrma) [inline os/drivers/wireless/virtual/Make.defs:24]: extra lines

File: os/include/tinyara/net/ioctl.h [hunk 17]
```diff
@@ -202,6 +202,7 @@
 #define SIOCGIFNAME      _SIOC(0x0056)  /* get active NIC name. */
 
 #define SIOCSLIPTYPE     _SIOC(0x0057)	/* Set IPv6 Address type */
+#define SIOCSLWNLEVT     _SIOC(0x0058)	/* Set IPv6 Address type */
```
Comment (sangam-swami) [inline os/include/tinyara/net/ioctl.h:205]: is the comment for event or type?

File: os/drivers/lwnl/lwnl_evt_queue.c [hunk 18]
```diff
@@ -22,24 +22,21 @@
...
 #include <debug.h>
 #include <net/if.h>
+#include <pthread.h>
 #include <tinyara/net/if/wifi.h>
 #include "lwnl_evt_queue.h"
+#include "lwnl_log.h"
 
-#define LWQ_LOCK
-#define LWQ_UNLOCK
+static sem_t g_wm_sem;
```
Comment (sijupunnnoose) [inline os/drivers/lwnl/lwnl_evt_queue.c:30]: locally many semaphores, can you please check do we need it really? Cant we make it as part of lwnl_event  or lwnl_queue struct instead? Also semaphore with same name is used in other file, anyway static. But confusing.

File: os/drivers/lwnl/lwnl_evt_queue.c [hunk 19]
```diff
@@ -181,14 +140,38 @@ static int _lwnl_copy_scan_info(char **buffer, trwifi_scan_list_s *scan_list)
...
+	LWQ_LOCK;
+	for (int i = 0; i < LWNL_NPOLLWAITERS; i++) {
+		g_queue[i].filep = NULL;
+		g_queue[i].front = NULL;
+		g_queue[i].rear = NULL;
+		g_queue[i].check_header = 0;
+	}
+	g_connected = 0;
+
+	int res = sem_init(&g_wm_sem, 0, 1);
```
Comment (sijupunnnoose) [inline os/drivers/lwnl/lwnl_evt_queue.c:167]: May I know, where is sem_destroy called?

**Review by sijupunnnoose — CHANGES_REQUESTED** [review]
Please recheck some of the neq changes especially the semaphore addition, lock/unlock, any deadlock, race-condition if any with the new changes,

File: os/drivers/lwnl/lwnl.c (hunk 10 above)
Comment (pillip8282) [reply os/drivers/lwnl/lwnl.c:365]: do you have any idea how to fix it?

File: apps/examples/wifi_manager_sample/wm_test_on_off.c (hunk 12 above)
Comment (pillip8282) [reply apps/examples/wifi_manager_sample/wm_test_on_off.c:25]: it calls semaphore

File: os/drivers/lwnl/lwnl_evt_queue.c (hunk 18 above)
Comment (pillip8282) [reply os/drivers/lwnl/lwnl_evt_queue.c:30]: yes those are necessary. g_wm_sem is to send signal to main task and g_queue_sem is to protect event queue.
And I don't understand how to use lwnl_event in here. only lwnl can access lwnl queue.
I changed the name of those semaphores. thanks

File: framework/src/wifi_manager/wifi_manager_cb.c [hunk 20]
```diff
@@ -111,32 +111,32 @@ void _handle_user_cb(_wifimgr_usr_cb_type_e evt, void *arg)
...
 			cbk->sta_connected(WIFI_MANAGER_FAIL);
 			break;
 		case CB_STA_DISCONNECTED:
-			WM_LOG_VERBOSE("[WM] call sta disconnect event\n");
+			WM_LOG_INFO("[WM] call sta disconnect event\n");
 			cbk->sta_disconnected(WIFI_MANAGER_DISCONNECT);
 			break;
 		case CB_STA_RECONNECTED:
-			WM_LOG_VERBOSE("[WM] call sta disconnect event\n");
+			WM_LOG_INFO("[WM] call sta disconnect event\n");
```
Comment (jeongchanKim) [inline framework/src/wifi_manager/wifi_manager_cb.c:127]: Is this RECONNECT? or DISCONNECT?
If reconnect, we'd better change the log to reconnect.

**Review by sijupunnnoose — APPROVED** [review]
Looks good me now, someone can review other changes and approve it. Thanks

**Review by jeongchanKim — APPROVED** [review]
LGTM

**Review by sangwon03 — APPROVED** [review]
LGTM
